content/posts/2019-03-job-search-2-parsing.md

```diff
@@ -0,0 +1,118 @@
+
+We used our hand-rolled parser until very recently, but as of this week we moved to using a [parser combinator](https://en.wikipedia.org/wiki/Parser_combinator) library called [Pidgin](https://github.com/benjamin-hodgson/Pidgin). This library is maintained by our very own parser wizard [Benjamin Hodgson](https://benjamin.pizza/) and eliminates a lot of the mistakes that are common in writing your own parser. It also performs very well compared to the hand-rolled version while being more maintainable; everybody wins!
+
+Originally I wrote this post about building the hand-rolled parser but trying to explain it concisely wound up being complex and verbose. I think this points to it being hard to reason about and maintain so I set out to write the parser combinator version instead. It turned out well enough (passes all tests and performs better) that I thought I'd write about that instead...
+
+Onwards, let's talk about the steps we took to build our parser!
+
+## Defining the Grammar
+
+We started with an idea of the language we wanted to implement. This is referred to as the grammar. It's common to use [Extended Backus-Naur Form (EBNF)](https://en.m.wikipedia.org/wiki/Extended_Backus–Naur_form)  which is a syntax used to desribe the individual components of the grammar. Here's a snippet of Jobs Query Language (JQL) in EBNF:
+
+```
+<and> ::= 'and' | '&&';
+<or> ::= 'or' | '||';
+<not> ::= 'not' | '-';
+<quote> ::= '"';
+<colon> ::= ':';
+<lparen> ::= '(';
+<rparen> ::= ')';
+<parens> ::= <lparen> | <rparen>;
+<lbracket> ::= '[';
+<rbracket> ::= ']';
+<brackets> ::= <lbracket> | <rbracket>;
+<string> ::= { <any_character> - (<brackets> | <parens> | <quote>) };
+<quoted_string> ::= <quote> { <any_character> - <quote> } <quote>;
+<tag> ::= { <letter> | <number> | '-' | '.' | '#' | '+' | '*' };
+<modifier> ::= { <letter> | '-' } <colon>;
+<range> ::= '.', '.';
+<unit> ::= <letter>, <letter>, <letter> | <letter>, <letter> | <letter>;
+<literal> ::= <number> | <string> | <quoted_string>;
+<term> ::= 
+    <modifier>, <colon>, <literal> |
+    <lbracket>, <tag>, <rbracket> | 
+    <string> | 
+    <quoted_string>;
+    
+<expression> ::= 
+    <lparen>, <expression>, <rparen> | 
+    <not>, <expression> |
+    <expression>, <and>, <expression> | 
+    <expression>, <or>, <expression> |
+    <term>;
+```
+
+This supports a wide variety of inputs for search:
+
+- **Simple Text**: `hello world`
+- **Quoted Text**: `"hello world"`
+- **Tags**: `[asp.net]`
+- **Modifiers**: `salary:10000` or `salary:10000USD`
+- **Ranges**: `salary:10000..20000` or `salary:..50000` or `salary:10000..`
+- **Expressions**: `[asp.net] or "hello world"` or `[c#] and not [java] and salary:10000..20000`
+- **Complex Expressions**: `([asp.net] or "hello world") and (([c#] and not [java]) or salary:10000..)`
```

Comment:
I'd start with this plain-text description of the language before getting into the BNF. I'd even consider dropping the BNF altogether, it doesn't necessarily tell the casual reader more than the text does.